src/tools/rbd_mirror/image_map/Policy.h

```diff
@@ -0,0 +1,96 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_POLICY_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_POLICY_H
+
+#include <map>
+#include <vector>
+#include <tuple>
+
+#include "common/RWLock.h"
+#include "cls/rbd/cls_rbd_types.h"
+#include "include/rados/librados.hpp"
+
+namespace librbd { class ImageCtx; }
+
+namespace rbd {
+namespace mirror {
+
+template <typename> struct Threads;
+
+namespace image_map {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class Policy {
+public:
+  Policy(librados::IoCtx &ioctx);
+
+  virtual ~Policy() {
+  }
+
+  // build an initial instance map
+  void build_map(const std::map<std::string, cls::rbd::MirrorImageMap> &image_mapping);
+
+  // lookup an image from the map
+  std::string lookup(const std::string &global_image_id);
+
+  // map an image
+  std::string map(const std::string &global_image_id);
+
+  // unmap (remove) an image from the map
+  bool unmap(const std::string &global_image_id);
+
+  // shuffle images when instances are added/removed
+  void shuffle(const std::vector<std::string> &instances,
+               std::map<std::string, std::pair<std::string, std::string> > *remap_ids);
+
+  // is an instance_id tracked in the map
+  bool instance_present(const std::string &instance_id);
+
+  static const std::string UNMAPPED_INSTANCE_ID;
+
+private:
+  librados::IoCtx &m_ioctx;
+  std::map<std::string, utime_t> m_shuffled_timestamp;
+
+  std::string _lookup(const std::string &global_image_id);
```

Comment:
Nit: prefer just passing the lock as an argument to indicate that the lock is already held vice prefixing w/ ```_```